Make probes independent of leveling

thinkyhead opened the pull request:
This PR makes probing independent of leveling, so you can use your Z probe for homing even if you don't have bed leveling enabled.

Already merged in separate PRs:
- Move all deploy/stow handling into `deploy_z_probe` / `stow_z_probe`.
- Remove support for XY servo endstops (never properly supported anyway).
- Make `zprobe_zoffset` dependent only on `HAS_BED_PROBE`.
- Make `M851` standard with no option to remap it, and only based on `HAS_BED_PROBE`.
- Enable probing GCodes and probe-homing based only on `HAS_BED_PROBE`.
- Simplify stowing code at the end of `gcode_G29`. (Over-simplified?)
- Sanity check Allen Key Probe for grid leveling only if leveling is enabled.

And…
- Initialize `current_position` to `home_offset` on boot
- Show uncorrected position in `set_bed_level_equation_3pts`
- Comment on `run_z_probe`
- Never do deploy-and-stow in `probe_pt` for Allen Key or Sled probes
- Split up `endstop_move` functions for probes / endstops
- Move `deploy_z_probe` after the `setup_for_probe_or_endstop_move` call in `G29`
- Use `sync_plan_position_delta` where needed using a macro

Roxy-3D commented:
I think we need to do this.  But I have a question about whether it actually fixes the G28 Z case with a servo.   At the tail end of G28, we have this code:

```
    // Retract Servo endstop if enabled
    #if HAS_SERVO_ENDSTOPS
      if (_Z_SERVO_TEST && servo_endstop_id[axis] >= 0) {
        #if ENABLED(DEBUG_LEVELING_FEATURE)
          if (DEBUGGING(LEVELING)) SERIAL_ECHOLNPGM("> SERVO_ENDSTOPS > Stow with servo.move()");
        #endif
        servo[servo_endstop_id[axis]].move(servo_endstop_angle[axis][1]);
        if (_Z_PROBE_SUBTEST) endstops.enable_z_probe(false);
      }
    #endif
```

The code is going straight to the servo and not calling the stow_z_probe() function.  Don't we need to change it to call stow_z_probe() to get the desired effect?  Or alternatively, before this code messes with the servo, we need to call raise_z_after_probing() before it messes with the servo?

thinkyhead commented:
That code should only apply to XY servos in the case where you have a Z probe. The code above that block will have already called `stow_z_probe` in the case of a Z probe.

``` cpp
// Put away the Z probe with a function
#if ENABLED(Z_PROBE_SLED) || SERVO_LEVELING || ENABLED(FIX_MOUNTED_PROBE)
  if (axis == Z_AXIS && axis_home_dir < 0) {
    stow_z_probe();
  }
#endif
```

For a servo-based Z probe, the second block ends up being reduced to…

``` cpp
// Retract X, Y (or Z) Servo endstop if enabled
#if ENABLED(HAS_SERVO_ENDSTOPS)
  if (axis != Z_AXIS && SERVO_ENDSTOP_EXISTS(axis)) {
    STOW_SERVO_ENDSTOP(axis);
  }
#endif // HAS_SERVO_ENDSTOPS
```

Roxy-3D commented:
I see what you are saying...   But I put the `raise_z_after_probing()` call in that X&Y code and it fixed the problem.   My guess is it fell through that code first, did what I needed, and then later went through the Z code that was still wrong.   

Wow!

thinkyhead commented:
I'm unsure. With your addition, it should have reduced to…

``` cpp
// Retract Servo endstop if enabled
#if HAS_SERVO_ENDSTOPS
  if (axis != Z_AXIS && servo_endstop_id[axis] >= 0) {
    raise_z_after_probing();
    STOW_SERVO_ENDSTOP(axis);
  }
#endif
```

…which would only run if you have an index set for X or Y servos. Did you leave these commented out?:

``` cpp
//#define X_ENDSTOP_SERVO_NR 1
//#define Y_ENDSTOP_SERVO_NR 2
```

Blue-Marlin commented:
It seems to be the right time to (remind) you that `SERVO_ENDSTOPS` can be dropped at all if the probes would always be usable in G28. They are only defined when bed leveling is defined. That's a conceptual error.

Roxy-3D commented:
@Blue-Marlin  Can you expand a little bit on that thought?   It isn't quite coming across.   But from what I did understand, I think I would agree if a probe is defined, it should be usable within the G28 command.     I also think that G28 should not be concerned with how a probe does it's work.  If the probe is implemented with a servo that work should be done at a lower level.

Blue-Marlin commented:
Two stubs from `homeaxis()`:
https://github.com/MarlinFirmware/Marlin/blob/RCBugFix/Marlin/Marlin_main.cpp#L2284-L2302
https://github.com/MarlinFirmware/Marlin/blob/RCBugFix/Marlin/Marlin_main.cpp#L2416-L2451
In both of them the probe block (including `SERVO_LEVELING` what is the servo probe) is follower by the `HAS_SERVO_ENDSTOPS` block. That is necessary, because the probes are not defined without bed levelling. If the probes would be defined outside `AUTO_BED_LEVELING_FEATURE`
https://github.com/MarlinFirmware/Marlin/blob/RCBugFix/Marlin/Configuration.h#L553
https://github.com/MarlinFirmware/Marlin/blob/RCBugFix/Marlin/Configuration.h#L596-647
we could always use them for homing.
The same nesting exists in Marlin_main.cpp

What is done in `HAS_SERVO_ENDSTOPS` is (should be) just a rewrite of the probe deploy/stow code for the case the probes are absent. 
The advantage of eliminating the `HAS_SERVO_ENDSTOPS`-block and using the probes instead would be that **all** probes can be used for homing then. Ok, all probes you can deploy without knowing z - or you can deploy with not more risk then servo probes.

Roxy-3D commented:
Thank You for the explanation.   This sounds like something that should not be done prior to the Release Candidate going 'Golden'.  But we should try to do this as soon as we can after that happens.

thinkyhead commented:
I'm not sure if we're all clear about these flags. So a quick review…
- `HAS_SERVO_ENDSTOPS` is set whenever there is an X, Y, or Z servo endstop.
- `HAS_SERVO_ENDSTOPS` will always be set when `SERVO_LEVELING` is set.
- `SERVO_LEVELING` == `AUTO_BED_LEVELING_FEATURE && HAS_Z_ENDSTOP_SERVO`.
- The first block handling `SERVO_LEVELING` takes care of your Z Endstop Servo (servo probe) (when `AUTO_BED_LEVELING_FEATURE` is enabled).
- The second block handling `HAS_SERVO_ENDSTOPS` takes care of your X and Y Endstop Servos.
  - If `AUTO_BED_LEVELING_FEATURE` is disabled, then this block also takes care of your Z Endstop Servo, if there is one.

Yep, Marlin supports X and Y Servo Endstops (and Z Servo Endstops that are just endstops).

AnHardt commented:
No, it does not anymore. Take a look into `Endstops::update()`. There are relics in the servo angles array and some other places, but i don't expect them to work. By the way - they are completely useless and would only complicate the code.
Even if there would be working x/y servo endstops it would be time to drop them.

thinkyhead commented:
Happy to see this code go, then.

And why do we still have a three-element array for endstop servo angles, if we will only ever care about Z? Does this mean we could get rid of the other two elements now?

``` cpp
#define SERVO_ENDSTOP_ANGLES {{0,0}, {0,0}, {70,0}}
```

…replaced with…

``` cpp
#define Z_SERVO_ANGLES {70,0}
```

AnHardt commented:
>  Does this mean we could get rid of the other two elements now?

Yes.

AnHardt commented:
For this extruder with a servo we need again two defined angles. Maybe we can reuse the array with an other meaning.
However, the servo setup could be considerably simplified. What means - not in RC.

thinkyhead commented:
> For this extruder with a servo we need again two defined angles. Maybe we can reuse the array with an other meaning.

No need to re-use. It can just have its own distinct 2-member array.

thinkyhead commented:
Ok, I've inserted a commit removing support for the obsolete servo endstops, along with some sanity checks…

``` cpp
#elif defined(SERVO_ENDSTOP_ANGLES)
  #error "SERVO_ENDSTOP_ANGLES is deprecated. Use Z_SERVO_ANGLES instead."
#elif defined(X_ENDSTOP_SERVO_NR) || defined(Y_ENDSTOP_SERVO_NR)
  #error "X_ENDSTOP_SERVO_NR and Y_ENDSTOP_SERVO_NR are deprecated and should be removed."
#endif
```

And, now I've added some more commits to allow probing if you have a probe, not requiring ABL.

thinkyhead commented:
This PR has turned into a bigger cleanup than originally planned. But probing should be independent of ABL, and it turns out this isn't too hard to accomplish. That's the most major change here. Combining deploy/stow with `dock_sled` is the other "obvious" improvement. Still poring through to see where Travis is catching oversights.

thinkyhead committed:
Initialize current_position to home_offset on boot

thinkyhead committed:
Show uncorrected position in set_bed_level_equation_3pts

thinkyhead committed:
Make M401/M402 available with HAS_BED_PROBE

thinkyhead committed:
For all probes, at the end of G29 simply raise and stow

thinkyhead committed:
Comment on run_z_probe

thinkyhead committed:
Simpler handling of sled/allen-key in probe_pt

thinkyhead committed:
Split up endstop_move functions

thinkyhead committed:
Move probe deploy after setup call in G29

thinkyhead committed:
Use sync_plan_position_delta where needed

thinkyhead committed:
Cleanup debug in G28 and G29

thinkyhead committed:
Use a macro for kinematic sync_plan_position

thinkyhead committed:
Screen out `G29 E` for sled and allen key, allow single probe

thinkyhead committed:
Tweak some feedrate vars

AnHardt commented:
Great! Seems to work with my setups.

Now it becomes obvious. MECHANICAL_PROBE == FIX_MOUNTED_PROBE      
-   

```
...\sketch\Marlin_main.cpp:1613:15: warning: 'void clean_up_after_endstop_move()' defined but not used [-Wunused-function]

```

At the end of G28

```
  #if ENABLED(ENDSTOPS_ONLY_FOR_HOMING) // We don't need this condition anymore.
                                        // endstops.enable(false); knows how to do it.       
    endstops.enable(false);                                     // !!!
    #if ENABLED(DEBUG_LEVELING_FEATURE)
      if (DEBUGGING(LEVELING)) {
        SERIAL_ECHOLNPGM("ENDSTOPS_ONLY_FOR_HOMING endstops.enable(false)");
      }
    #endif
  #endif

  ...

    #endif

  feedrate = saved_feedrate;                                   // !!!
  feedrate_multiplier = saved_feedrate_multiplier;             // !!!
  refresh_cmd_timeout();                                       // !!!
  endstops.hit_on_purpose(); // clear endstop hit flags

  #if ENABLED(DEBUG_LEVELING_FEATURE)
```

This is the currently unused `clean_up_after_endstop_move()`

`clean_up_after_endstop_move()` can't be dependent on `HAS_BED_PROBE`

I can find `clean_up_after_endstop_or_probe_move();` two times in G29. It makes most sense at the very end, after the last stow move.

In M48 `setup_for_endstop_or_probe_move();` should be before the first move - some lines up. After the first probe we are to(o) deep for moving (star/circle). A raise is needed.

I wonder if that saving and restoring isn't better done locally in the currently calling function. If accidentally nested we get chaos with the globals. I hardly dare to say it, but macros could be right here.
*

```
    #endif // !DELTA

    // Final raise of Z axis after probing.
    raise_z_after_probing();                                   // This seems to be to much.

    // Stow the probe. Servo will raise if needed.
    stow_z_probe();

    #ifdef Z_PROBE_END_SCRIPT
```

AnHardt commented:
`Z_MIN_PROBE_REPEATABILITY_TEST` without `AUTO_BED_LEVELING_FEATURE` still has some problems to compile.

thinkyhead committed:
Make Z_RAISE_BETWEEN_PROBINGS a more general option

thinkyhead committed:
Make sure NUM_SERVOS is set for Z servo endstop

thinkyhead committed:
Add HAS_PROBING_PROCEDURE conditional

thinkyhead committed:
Add a dependency for z servo macros

thinkyhead committed:
Fix compile error for bed matrix in M48

thinkyhead commented:
I've added commits to allow compiling of `M48` by itself.

I'm also applying suggestions from @Roxy-3D.

The additional Z raise after probing (and Z raise in-general) I want to try fixing by having the Z raise go to an absolute Z position, so instead of…

```
current_position[Z_AXIS] + Z_RAISE_AFTER_PROBING
```

…we would simply use the equivalent of…

```
max(current_position[Z_AXIS], home_offset[Z_AXIS] + Z_RAISE_AFTER_PROBING)
```

thinkyhead committed:
Simpler cleanup in G28 / G29

thinkyhead committed:
Generalized probe raise, use for after probing

AnHardt commented:
The original idea about `Z_RAISE_AFTER_PROBING` was to have enough room to stow - not be at a hight after beeing ready. So  `Z_RAISE_AFTER_PROBING` and `Z_RAISE_BEFORE_PROBING` can be the same. If we want to have a minimal height after, this should be `MIN_Z_HEIGHT_FOR_HOMING` or its unification with `Z_RAISE_BETWEEN_PROBINGS`.

thinkyhead commented:
> The original idea about Z_RAISE_AFTER_PROBING was to have enough room to stow

I see it still serves that purpose. At least, it is used to raise for stowing a servo probe. But it is not being used as the raise for stowing in all cases, in part because there could be a double-raise because of how the `raise_after_probing` function is implemented.

The `probe_pt` function adds some "complications" to be able to do its thing:
- Raise to the given (absolute) Z coordinate before doing anything
- Move to the given XY - the probe point
- If flagged, `deploy_z_probe` (might raise)
- Call `run_z_probe` to probe the bed, remember the Z
- If flagged, `stow_z_probe` (might raise)
- Return the measured (not current) Z position

I agree that converting the "Z_RAISE" options to "Z_HEIGHT" options makes things better. But we could even leave the names alone and just always use them in exactly the same manner:

``` cpp
float z_dest = home_offset[Z_AXIS] + negative_zprobe_zoffset + Z_RAISE;
if (z_dest > current_position[Z_AXIS]) do_blocking_move_to_z(z_dest);
```

To make this viable everywhere we have to ensure that when the bed is probed, the "zero-point" where the probe is triggered is close to `home_offset[Z_AXIS] + zprobe_zoffset` – as opposed to zero.

Roxy-3D commented:
> I agree that converting the "Z_RAISE" options to "Z_HEIGHT" options makes things better. But we could even leave the names alone and just always use them in exactly the same manner:

I think we want to move to a Z_HEIGHT mindset.  That makes it much less likely to be at the wrong height (because we did multiple Z_RAISE_???).    And it makes it simpler to write resilient code that does the right thing.     Remember, the whole reason we have so many Z_RAISE_??? options right now is because some servo based Z-Probes needed extra room to retract and deploy.   Well, it made sense to do a RAISE to facilitate that.  But then more and more special cases came about and we kept going further and further down the Z_RAISE_??? path.

What we really need in all of the Z_RAISE_??? cases is a specified Z_HEIGHT_??? that is safe.    I think we should head in that directions.    But could we do it for the time being with the existing Z_RAISE_??? names?   

As soon as we go 'Golden' we could make the first point release nothing but all the global names we want to change.  That will make it much easier for people to stay in synch with the 'Golden' release and keep thier individual branches up to date.    The big problem with these sweeping global name changes is they create so many individual differences that all have to be looked at when you are merging code.   If I knew the first point release was nothing but name changes, I could get my code merged with the 'Stable Release'.   I can then accept the 500 differences for the first point release with confidence and not have to go through them line by line.

thinkyhead commented:
Currently the auto-leveling `G29` process uses `probe_pt` and has to work around it a bit.
The flow of `G29` now looks like this:
- Interpret `G29` parameters, set flags, etc.
- Call `setup_for_endstop_or_probe_move`
- Deploy the probe in the current position.
- Set feedrate to the Z homing feedrate (should do this earlier, or should not be needed)
- For grid probing, at each `probe_pt` decide before-height, whether ("`E`") to deploy/stow
- For 3-point probing, a simplified version of the above
- Calculate a matrix, correct the current position
- Call `raise_z_after_probing` (currently always raise)
- Call `stow_z_probe` (raise for servo)
- Call `clean_up_after_endstop_or_probe_move`

Note that the raise functions and `probe_pt` don't set/restore the feedrate. Some functions, like `run_z_probe` do set and restore it. More functions should set/restore feedrate for simplicity. The `probe_pt` function should definitely set/restore feedrate separately for XY and Z moves. We can review where feed rate is being set throughout the code and see if we can generalize the rates that go with certain moves, ensure they are always set correctly.

Roxy-3D commented:
Obviously, `run_z_probe()` needs to mess with the `feedrate`. But other than that, do any of the other functions that use the `Z_RAISE_???` values need to adjust `feedrate`?  It might be simpler if none of them messed with it.

thinkyhead committed:
Save and restore feedrate in more places

AnHardt commented:
If the shitty raises for the servo probe would be completely in the deploy/stow-functions we could deploy at the beginning for the case of not 'E'. When the probe knows if it is deployed or not, we could deploy for every probe-point and retract if we have an 'E'. Behind the loop we could stow an other time. (Just an idea - will verify tomorrow.)

Changing functionality but leaving the old name would be a bit confusing. Wouldn't it?

thinkyhead commented:
@Roxy-3D With the latest commit I'm proposing that `feedrate` be treated with care by functions that need to modify it. Functions that alter it to do moves will restore it. And it is set to the appropriate feedrate for the given move. For Z moves it's set to `homing_feedrate[Z_AXIS]` and for XY moves it is set to `XY_PROBE_FEEDRATE`, a macro that differs depending on settings.

Roxy-3D commented:
> Functions that alter it to do moves will restore it.

I think this is an easy one to get agreement.   If a routine really needs to alter it, then it should put it back at the end!

thinkyhead commented:
> Just an idea - will verify tomorrow.

@AnHardt Makes sense to me. The probe knows if it has been deployed or not, so redundant calls to deploy or stow are harmless. The first call to `probe_pt` already takes advantage of this, so it won't deploy the already-deployed probe. ~~I'm looking at `G30` to see if it can use `probe_pt` and save a little code too….~~

AnHardt commented on Marlin/Marlin_main.cpp:
```diff
@@ -1712,8 +1714,12 @@ static void clean_up_after_endstop_or_probe_move() {
+      feedrate = homing_feedrate[Z_AXIS];
       do_blocking_move_to_z(z_dest);
+      feedrate = old_feedrate;
+    }
```
Do you want to do this for every call to `do_blocking_move_to_z()`? It should better be **in**  `do_blocking_move_to_z()`.  In fakt it's allreay in - in `do_blocking_move_to()`.

thinkyhead committed:
Pass a raise to probe_pt, not a Z position

thinkyhead committed:
Return Z position in run_z_probe

thinkyhead committed:
Simplify initial raise in M48

thinkyhead commented:
> MECHANICAL_PROBE == FIX_MOUNTED_PROBE

It seems to be true, so we could drop `MECHANICAL_PROBE` entirely, although the distinction might still be useful for users, even though it compiles the same code. We don't have an explicit `INDUCTIVE_PROBE` option – instead users must select `FIX_MOUNTED_PROBE`. If we drop `MECHANICAL_PROBE` then we need to make it clear that `FIX_MOUNTED_PROBE` applies to both inductive probes and probes that make use of the nozzle and/or pressure-switch (usually with positive Z probe offset).

We haven't any option for a probe that needs to be deployed and stowed by hand, requiring either a long pause at the start of probing, or waiting for LCD click. Do you think we should include support such a probe, with a name like `MANUALLY_DEPLOYED_PROBE`?

thinkyhead committed:
Goodbye MECHANICAL_PROBE

thinkyhead committed:
Apply some fixes from Andreas

Derived from
https://github.com/AnHardt/Marlin/commit/6e8ede8c694fa4d9e3c769840b09d56
7f8d8e0dd

Roxy-3D commented:
> We haven't any option for a probe that needs to be deployed and stowed by hand, requiring either a long pause at the start of probing, or waiting for LCD click. Do you think we should include support such a probe, with a name like MANUALLY_DEPLOYED_PROBE?

YES!  We definitely need that! I have a lame Geeetech printer that has a Manually Deployed Probe.   What they did in the Geeetech distributed software is move the effector halfway down the towers, and then do a horizontal move at a super slow feedrate.  That gives you time to deploy the probe.  And then they do the same thing to retract it.

But that is just asking for trouble.  I have my firmware calling lcd_click() to confirm the printer can continue after both the deploy and the stow.

thinkyhead committed:
Define Z_SAFE_HOMING point when Z_PROBE_SLED is activated alone

thinkyhead commented:
The more I hear about other workarounds the more I like re-using endstop switches for simple inputs. On many machines, at least 2, if not 3 endstop plugs are unused, so making them into assignable push-buttons would be a cheap way to add up to 3 simple functions to your old RepRap.

I suppose I could even warm up to the idea of using the endstops (when not homing) for other functions also, although until the endstops are clear of obstruction they are useless. You could wire NC switches in series, NO in parallel, but again, the endstops still need to be unpressed.

AnHardt commented:
@thinkyhead 

> `MANUALLY_DEPLOYED_PROBE`

I'm a bit afraid to say that's more a long therm project, because you usually immediately begin with the realization then. So in this case i'll try to say - yes a very good idea - we need it now. Maybe waiting will be easier for you then.

An other good thing would be a `MANUAL_PROBE`. The deploy ends at `MESH_HOME_SEARCH_Z`. The actual probing is done by the click-wheel of the panel and ends with a click.  Pretty much like today's manual probing in mesh-bed-leveling.

Other projects could include  
making Allen-Key-Probes fit for Cartesian's. 
Modularisierung probes (check - sequence of moves - special action - move back sequence - check : all parts can be omitted)

> Buttons / Endstops

You kept me busy today.

thinkyhead commented:
> You kept me busy today.

I can tell we are all hungry for simplification in this area!

jbrazio commented:
I like @thinkyhead idea, a "endstop input" if it has a endstop connected it will trigger endstop_stuff() other wise it will trigger button_stuff(), dead simple with no PIN #undefs and other crazy stuff.

Roxy-3D commented:
> The more I hear about other workarounds the more I like re-using endstop switches for simple inputs. On many machines, at least 2, if not 3 endstop plugs are unused, so making them into assignable push-buttons would be a cheap way to add up to 3 simple functions to your old RepRap.

I think this will help a lot of people.

> > MANUALLY_DEPLOYED_PROBE
> > I'm a bit afraid to say that's more a long therm project, because you usually immediately begin with the realization then. So in this case i'll try to say - yes a very good idea - we need it now. Maybe waiting will be easier for you then.

I believe this is very low risk and very easy.  If the MANUALLY_DEPLOYED_PROBE is enabled, we just wait for an lcd_clicked() to continue in both stow and deploy of a probe.   Later, we could do the extra work to make it flexible and use the CONTINUE_FEATURE.

> An other good thing would be a MANUAL_PROBE. The deploy ends at MESH_HOME_SEARCH_Z. The actual probing is done by the click-wheel of the panel and ends with a click. Pretty much like today's manual probing in mesh-bed-leveling.   

Yes!  And it is already coded and working in the Unified Bed Leveling code:

```

float use_encoder_wheel_to_measure_point() {
    G29_has_control_of_LCD_Panel++;
    while (!G29_lcd_clicked() ) {       // we need the loop to move the nozzle based on the encoder wheel here!
        idle();
        if ( G29_encoderDiff != 0) {
            float new_z;                    
            // We define a new variable so we can know ahead of time where we are trying to go.
            // The reason is we want G29_encoderDiff cleared so an interrupt can update it even before the move 
            // is complete.  (So the dial feels responsive to user)
            new_z = current_position[Z_AXIS]+((float) G29_encoderDiff)/100.0;
            G29_encoderDiff = 0;            
            do_blocking_move_to_z( new_z );
        }   
    }
    while ( G29_lcd_clicked() ) {       // debounce and wait
        idle();
    }
    G29_has_control_of_LCD_Panel--;
    return current_position[Z_AXIS];
}
```

thinkyhead commented:
> Modularisierung probes

Very much so. I see a possibility of merger, top-down at first, to add automated MBL, and to generalize Grid and 3-point leveling. (In the direction of Universal Bed Leveling.) Basically the `G29` function can be reduced to…
- Enter the `G29` procedure for Automated probing
- Parse arguments such as grid size
  - For mesh probing set the mesh size, initializing mbl
  - For grid probing, use the size for the procedure
- Get the probe ready (if you need to do it early, such as for a sled)
- Multiply X grid points by Y grid points for the loop size
- Loop through the points…
  - Ask the "grid object" for the XY of the indexed probe point
    - The grid object will manage zigzag of points
  - Probe the XY to get Z
  - Tell the grid object the Z for the loop index (or xy indexes, or xy points)
- Put away the probe, if needed.
- Tell the grid object to calculate, enable compensation, etc.

The  `mesh_bed_leveling` object is a pretty good pattern to follow for the grid and three-point leveling objects, as far as how it acts as a data provider to a leveling procedure rather than doing any leveling actions itself. The code in `G29` will become a lot easier to follow as these parts are offloaded into class methods.

thinkyhead commented:
> use_encoder_wheel_to_measure_point

👍

Roxy-3D commented:
> The mesh_bed_leveling object is a pretty good pattern to follow for the grid and three-point leveling objects, as far as how it acts as a data provider to a leveling procedure rather than doing any leveling actions itself. The code in G29 will become a lot easier to follow as these parts are offloaded into class methods.

Please hold off on doing anything with the Mesh object or adding 3-Point or Grid or Delta objects.   What I've got cooking already embraces all of those.   And all of the nested #ifdef's are gone.

AnHardt commented:
@Roxy-3D 
If we'd have a preview we wouldn't have to guess.

thinkyhead commented:
There are enough conceptual changes (mainly in `probe_pt` taking an offset rather than a Z position) that this PR probably should get another test run. All the commands that apply… `G28`, `G29`, `G30`, `G31`, `G32`, `M48`, `M401`, `M402`… and any others you might think of. The code considers Z raises as being above the probe's Z trigger position, and in other places the Z raise has been made double-raise-proof.

Roxy-3D commented:
@AnHardt

> If we'd have a preview we wouldn't have to guess.

How about in a couple of weeks?   Basically, it is a high resolution (15x15) Mesh system that allows 3-Point and Grid Based to tilt the Mesh.  Right now, I'm working on the Mesh editor to make it easy to update and/or fine tune your Mesh's.   And it doesn't force you to prepare a Mesh.  If you have a sufficiently flat piece of glass, you can just tell it to zero the entire Mesh and perform a tilt based on 3-Point or Grid probing.  (So it is backwards compatible).

I don't have everything working...  And my coding style is kind of chaotic where I paste old functions and header files into my files just so I have easy reference.   I have to do some clean up before I'm ready to share the code.    

What I'm trying to do is stay synchronized with ThinkyHead so when we promote the Release Candidate to 'Golden' status, we have a Development Branch ready to go with Unified Bed Leveling based on the Stable Branch.

AnHardt commented:
@thinkyhead 
I'll try to test as much as i can, but i'v to prepare a birthday party an guests over the weekend.

@Roxy-3D 
You are underestimating us - we are able read Marlin.

Roxy-3D commented:
> You underestimate us - we can read Marlin.

No.  You are wrong.  I'm not under estimating you.  I know your code is clean and I don't want to share anything that is not just as clean.

AnHardt commented:
@Roxy-3D 

```
@)>--->--- (rose)
```

I guess @thinkyhead has an other opinion about my style. 
(However 3:55 am - good night)

thinkyhead commented:
This addresses enough bugs, logical errors, and especially compile issues, that I thought it best to merge it now. It's modular enough that we can backtrack on any section, should there be any unexpected behavior.

AnHardt commented:
Tried some things we talked about yesterday https://github.com/AnHardt/Marlin/pull/48

thinkyhead commented:
PR #4134 incorporates most of those changes, which are very sensible.
